### PR TITLE
Loading Spinner Visual Element for foundation

### PR DESCRIPTION
**Pull request by alexey-yaremenko**

### CODE REVIEW

**Comment by alexey-yaremenko** [comment]
Forgot to add comments to public endpoints.
gonna add in following commit

File: Runtime/UIElements/Controls/LoadingSpinner.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            Started = true;
+            UIUtils.SetElementDisplay(this, true);
+        }
+
+        public void Stop()
+        {
+            if (!Started)
+                return;
+
+            EditorApplication.update -= UpdateProgress;
```
Comment (stan-osipov) [inline Runtime/UIElements/Controls/LoadingSpinner.cs:68]: This would limit us to use it in Editor only.
Let's use  `schedule.Execute(UpdateProgress).Every(1);` instead

File: Runtime/UIElements/Controls/LoadingSpinner.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+            Started = false;
+            UIUtils.SetElementDisplay(this, false);
```
Comment (stan-osipov) [inline Runtime/UIElements/Controls/LoadingSpinner.cs:71]: Why `Display.None` is enough?

File: Runtime/UIElements/Controls/UIUtils.cs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+using UnityEngine.UIElements;
+
+static class UIUtils
+    {
+        static readonly string[] s_SizeUnits = { "KB", "MB", "GB", "TB" };
+
+        public static void SetElementDisplay(VisualElement element, bool value)
```
Comment (stan-osipov) [inline Runtime/UIElements/Controls/UIUtils.cs:11]: I am not sure why do we need this.
If we have `DisplayStyle.None` there is no point to also apply `Visibility.Hidden`

I might be missing something tho.

File: Runtime/UIElements/Controls/UIUtils.cs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        {
+            if (element == null)
+                return;
+
+            element.style.display = value ? DisplayStyle.Flex : DisplayStyle.None;
+            element.style.visibility = value ? Visibility.Visible : Visibility.Hidden;
+        }
+
+        public static void SetElementDisplayNonEmpty(TextElement element)
```
Comment (stan-osipov) [inline Runtime/UIElements/Controls/UIUtils.cs:20]: I also not sure why to have another method for this.
Plus method semantics does not really tell me what the method would do.

File: Runtime/UIElements/Controls/UIUtils.cs [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        {
+            if (element == null)
+                return;
+
+            var nonEmpty = !String.IsNullOrEmpty(element.text);
+            element.style.display = nonEmpty ? DisplayStyle.Flex : DisplayStyle.None;
+            element.style.visibility = nonEmpty ? Visibility.Visible : Visibility.Hidden;
+        }
+
+        public static bool IsElementVisible(VisualElement element)
```
Comment (stan-osipov) [inline Runtime/UIElements/Controls/UIUtils.cs:30]: Same comment about method semantics. The fact that we have `Visibility.Visible` and `DisplayStyle.Flex` does not mean that element is actually "Visible"

File: Runtime/UIElements/Controls/UIUtils.cs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        {
+            return element.resolvedStyle.visibility == Visibility.Visible && element.resolvedStyle.display != DisplayStyle.None;
+        }
+
+        public static void ScrollIfNeeded(ScrollView container, VisualElement target)
```
Comment (stan-osipov) [inline Runtime/UIElements/Controls/UIUtils.cs:35]: `container.ScrollTo();` ? 
What's the difference?

File: Runtime/UIElements/Controls/UIUtils.cs [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+                container.scrollOffset = scroll;
+            }
+            else if (itemMaxY > maxY)
+            {
+                scroll.y += itemMaxY - maxY;
+                container.scrollOffset = scroll;
+            }
+        }
+
+        public static IEnumerable<T> GetParentsOfType<T>(VisualElement element) where T : VisualElement
```
Comment (stan-osipov) [inline Runtime/UIElements/Controls/UIUtils.cs:62]: Nice one, but I think it should go to the VisualElement extension methods:
We even have a PR for that in-flight: https://github.com/StansAssets/com.stansassets.foundation/pull/15

File: Runtime/UIElements/Controls/UIUtils.cs [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+                if (parent is T selected)
+                    result.Add(selected);
+
+                parent = parent.parent;
+            }
+
+            return result;
+        }
+
+        public static T GetParentOfType<T>(VisualElement element) where T : VisualElement
```
Comment (stan-osipov) [inline Runtime/UIElements/Controls/UIUtils.cs:78]: same as above.

Comment (stan-osipov) [reply Runtime/UIElements/Controls/UIUtils.cs:78]: Also maybe rename to GetFirstParentOfType
And another idea is to actually add this extension to `UQueryBuilder` -> if that is possible this would be the preferred way to go.

P.S. Very needed utility method!

File: Runtime/UIElements/Controls/UIUtils.cs [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+        {
+            return GetParentsOfType<T>(element).FirstOrDefault();
+        }
+
+        public static string ConvertToHumanReadableSize(ulong sizeInBytes)
```
Comment (stan-osipov) [inline Runtime/UIElements/Controls/UIUtils.cs:83]: Is this UI related? Feels more like `NameUtilities`
Or probably something that is only related to the specific package not the foundation.

File: Runtime/UIElements/Controls/UIUtils.cs [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+            var order = 0;
+            while (len >= 1024 && order < s_SizeUnits.Length - 1)
+            {
+                order++;
+                len = len / 1024;
+            }
+            return $"{len:0.##} {s_SizeUnits[order]}";
+        }
+
+        public static void ShowTextTooltipOnSizeChange<T>(this T element, int deltaWidth = 0) where T : TextElement
```
Comment (stan-osipov) [inline Runtime/UIElements/Controls/UIUtils.cs:95]: Again feel like something very specific and not related to the foundation lib.
I can't imagine wide use case of that

**Comment by alexey-yaremenko** [comment]
- comments added to public endpoints
- now Loading Spinner is ready to be used at runtime, not editor-only (editor update replaced with schedule.Execute)
- got rid of UIUtils class. Parent queries will be added separately with another PR after we merge #15 into master

File: Runtime/UIElements/Controls/LoadingSpinner.cs [hunk 11]
```diff
@@ -0,0 +1,90 @@
+﻿using JetBrains.Annotations;
+using UnityEngine;
+using UnityEngine.UIElements;
+
+namespace StansAssets.Foundation.UIElements
+{
+    /// <summary>
+    /// The LoadingSpinner control. let's you place buttons group with the labels or images
```
Comment (stan-osipov) [inline Runtime/UIElements/Controls/LoadingSpinner.cs:8]: let's you place buttons group with the labels or images?

File: Runtime/UIElements/Controls/LoadingSpinner.cs [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    /// </summary>
+    public sealed class LoadingSpinner : VisualElement
+    {
+        [UsedImplicitly]
+        internal new class UxmlFactory : UxmlFactory<LoadingSpinner, UxmlTraits> { }
+
+        internal new class UxmlTraits : BindableElement.UxmlTraits { }
+
+        /// <summary>
+        /// Returns true if control is visible and updating it's progress
```
Comment (stan-osipov) [inline Runtime/UIElements/Controls/LoadingSpinner.cs:18]: A bit confusing since it does not have progress.

File: Runtime/UIElements/Controls/LoadingSpinner.cs [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+        /// <summary>
+        /// Creates LoadingSpinner control
+        /// </summary>
+        public LoadingSpinner()
+        {
+            styleSheets.Add( Resources.Load<StyleSheet>("LoadingSpinner"));
+            AddToClassList(UssClassName);
+
+            Started = false;
+            style.display = DisplayStyle.None;
```
Comment (stan-osipov) [inline Runtime/UIElements/Controls/LoadingSpinner.cs:42]: This feels like something that may create an issue when using the UI Builder.

I am not sure why do we need all the additional API like Start / Stop. I can't imagine the case when you would need "stopped" but visible loading spinner.

Suggestion: Let's start the spinner when VE is attached to the planned and stop the animation when it's deattached.
And let the user to pick show/hide approach by himself.
For example as the user I can do my show hide spinner logic with:

- Add / Remove from the panel
- `style.display`
- `style.visibility`

File: Runtime/UIElements/Controls/LoadingSpinner.cs (hunk 11 above)
Comment (alexey-yaremenko) [reply Runtime/UIElements/Controls/LoadingSpinner.cs:8]: no, it's just a spinner. you can embed it in any control you want

**Comment by alexey-yaremenko** [comment]
Loading Spinner doesn't override style by itself, it's up to the user now.
and it uses AttachTo/DetachFrom PanelEvents for it's internal activate/deactivate logic
